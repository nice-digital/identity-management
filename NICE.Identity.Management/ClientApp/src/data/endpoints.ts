const API_BASE_URL = process.env.REACT_APP_API_BASE_URL
	? process.env.REACT_APP_API_BASE_URL
	: "#{REACT_APP_API_BASE_URL}";

export const Endpoints = {
	usersList: `${API_BASE_URL}/users`, // get all users
	user: (userId: string | number): string => `${API_BASE_URL}/users/${userId}`,

	servicesList: `${API_BASE_URL}/services`,
	service: (serviceId: string | number): string =>
		`${API_BASE_URL}/services/${serviceId}`,
	
	userRolesByWebsite: (userId: string | number, websiteId: string | number): string =>
		`${API_BASE_URL}/users/${userId}/rolesbywebsite/${websiteId}`,

	verificationEmail: `${API_BASE_URL}/VerificationEmail/VerificationEmail`,
	websitesList: `${API_BASE_URL}/websites`,
};
