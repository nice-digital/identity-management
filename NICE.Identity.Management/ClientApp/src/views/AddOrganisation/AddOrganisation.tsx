import React, { Component } from "react";

export class AddOrganisation extends Component {
	render(): JSX.Element {
		return (
			<div>
				<p>Someting here</p>
			</div>
		);
	}
}
