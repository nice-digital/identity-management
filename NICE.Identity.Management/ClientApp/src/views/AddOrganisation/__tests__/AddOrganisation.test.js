import React from "react";
import { act } from 'react-dom/test-utils';
import { render, fireEvent, screen } from "@testing-library/react";
import { rest } from "msw";
import { setupServer } from "msw/node";
import { MemoryRouter } from "react-router";
import userEvent from "@testing-library/user-event";
import { AddOrganisation } from "../AddOrganisation";
import { Endpoints } from "src/data/endpoints";
import organisations from "./organisations.json";

const consoleErrorReset = console.error;

beforeEach(() => {
	console.error = consoleErrorReset;
});

const server = setupServer(
  rest.get(Endpoints.organisationsList, (req, res, ctx) => {
    return res(ctx.json([]));
  }),
  rest.post(Endpoints.organisationsList, (req, res, ctx) => {
    return res(ctx.json({}));
  })
);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

test("should match the snapshot on load", () => {
  const {container} = render(<AddOrganisation />, {wrapper: MemoryRouter});
	expect(container).toMatchSnapshot();
});

test("should disable form submit button when clicked", async () => {
  jest.useFakeTimers();
  render(<AddOrganisation />, {wrapper: MemoryRouter});
  const orgNameInput = screen.getByLabelText("Organisation name");
  let submitButton = screen.getByText("Save organisation");
  await act(async () => {
    fireEvent.change(orgNameInput, {target: {value: "Org Ninety Nine"}});
    jest.advanceTimersByTime(1000);
  });
  userEvent.click(submitButton);
  submitButton = await screen.findByText("Loading...");
  expect(submitButton).toBeInTheDocument();
  expect(submitButton).toBeDisabled();
  jest.useRealTimers();
});

test("should display confirmation message once fetchData post is successfully complete", async () => {
  jest.useFakeTimers();
  render(<AddOrganisation />, {wrapper: MemoryRouter});
  const orgNameInput = screen.getByLabelText("Organisation name");
  const submitButton = screen.getByText("Save organisation");
  await act(async () => {
    fireEvent.change(orgNameInput, {target: {value: "Org Ninety Nine"}});
  });
  userEvent.click(submitButton);
  const orgAddedMessage = await screen.findByText("New organisation has been added successfully.");
  expect(orgAddedMessage).toBeInTheDocument();
  expect(screen.getByRole("status")).toBeInTheDocument();
  jest.useRealTimers();
});

test("should show error message when fetchData post fails", async () => {
  jest.useFakeTimers();
  console.error = jest.fn();
  server.use(
		rest.post(Endpoints.organisationsList, (req, res, ctx) => {
			return res.once(
				ctx.status(500),
				ctx.json({})
			);
		})
	);
  render(<AddOrganisation />, {wrapper: MemoryRouter});
  const orgNameInput = screen.getByLabelText("Organisation name");
  const submitButton = screen.getByText("Save organisation");
  await act(async () => {
    fireEvent.change(orgNameInput, {target: {value: "Org Ninety Nine"}});
  });
  userEvent.click(submitButton);
  const errorAlert = await screen.findByRole("alert");
  expect(errorAlert).toBeInTheDocument();
  jest.useRealTimers();
});

test("should show validation error when name is invalid format", async () => {
  jest.useFakeTimers();
  render(<MemoryRouter><AddOrganisation /></MemoryRouter>);
  const orgNameInput = screen.getByLabelText("Organisation name");
  const user = userEvent.setup();
  orgNameInput.focus();
  await act(async () => {
    user.type(orgNameInput, 'a');
  });
  user.tab();
  const orgNotFoundMessage = await screen.findByText("Organisation name should be alphanumeric and be between 2-100 characters");
  expect(screen.getByText("Save organisation")).toHaveFocus();
  expect(orgNotFoundMessage).toBeInTheDocument();
  expect(orgNameInput).toBeInvalid();
  jest.useRealTimers();
});

test("should show validation error when org name is in use already", async () => {
  jest.useFakeTimers();
  server.use(
    rest.get(Endpoints.organisationsList, (req, res, ctx) => {
      return res.once(ctx.json(organisations));
    })
  );
  const user = userEvent.setup();
  render(<AddOrganisation />, {wrapper: MemoryRouter});
  const orgNameInput = screen.getByLabelText("Organisation name");
  orgNameInput.focus();
  await act(async () => {
    //user.paste("Org 1");
    fireEvent.change(orgNameInput, {target: {value: "Org 1"}});
    jest.advanceTimersByTime(1000);
  });   
  user.tab();
  const orgInUseMessage = await screen.findByText("Cannot add Org 1, that organisation already exists!");
  expect(orgInUseMessage).toBeInTheDocument();
  jest.useRealTimers();
});

