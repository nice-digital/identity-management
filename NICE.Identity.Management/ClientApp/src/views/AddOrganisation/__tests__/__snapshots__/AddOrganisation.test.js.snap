// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddOrganisation should match the snapshot on load 1`] = `
<Fragment>
  <Breadcrumbs>
    <Breadcrumb
      elementType={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "propTypes": Object {
            "innerRef": [Function],
            "onClick": [Function],
            "replace": [Function],
            "target": [Function],
            "to": [Function],
          },
          "render": [Function],
        }
      }
      to="/overview"
    >
      Administration
    </Breadcrumb>
    <Breadcrumb
      elementType={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "propTypes": Object {
            "innerRef": [Function],
            "onClick": [Function],
            "replace": [Function],
            "target": [Function],
            "to": [Function],
          },
          "render": [Function],
        }
      }
      to="/organisations"
    >
      Organisations
    </Breadcrumb>
    <Breadcrumb>
      Add organisation
    </Breadcrumb>
  </Breadcrumbs>
  <PageHeader
    heading="Add organisation"
  />
  <Grid
    elementType="div"
    equalHeight={false}
    gutter="standard"
    horizontalAlignment="left"
    reverse={false}
    verticalAlignment="top"
  >
    <GridItem
      cols={12}
      lg={4}
      md={6}
    >
      <form
        noValidate={true}
        onSubmit={[Function]}
      >
        <Input
          autoComplete="off"
          className="mb--e"
          data-qa-sel="name-input-add-organisation"
          disabled={false}
          error={false}
          errorMessage="Organisation name should be alphanumeric and be between 2-100 characters"
          label="Organisation name"
          maxLength="100"
          minLength="2"
          name="orgName"
          onBlur={[Function]}
          onChange={[Function]}
          pattern="^((?![<>]).)*$"
          placeholder="Add organisation name"
          required={true}
          type="text"
          value=""
        />
        <Button
          data-qa-sel="save-button-add-organisation"
          disabled={false}
          type="submit"
          variant="cta"
        >
          Save organisation
        </Button>
        <Button
          disabled={false}
          elementType={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "propTypes": Object {
                "innerRef": [Function],
                "onClick": [Function],
                "replace": [Function],
                "target": [Function],
                "to": [Function],
              },
              "render": [Function],
            }
          }
          to="/organisations"
          variant="secondary"
        >
          Cancel
        </Button>
      </form>
    </GridItem>
  </Grid>
</Fragment>
`;
